### PR TITLE
feat(container query): add size observer component

### PR DESCRIPTION
**Pull request by jessemartin**
This introduces a SizeObserver component, which allows media-query like rendering of a component
based on its height and width.

## Pull request checklist

* [x] Component `README.md` file is up-to-date.
* [x] Component is unit tested.

### CODE REVIEW

File: src/components/ContainerQuery/ContainerQuery.tsx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  target: HTMLElement;
+}
+
+type ResizeObserver = new (cb: (entries: ResizeObserverEntry[]) => void) => {
+  disconnect(): void;
+  observe(): void;
+  unobserve(): void;
+};
+
+export interface ContainerQueryProps extends NestableBaseComponentProps, FocusableComponentProps {
```
Comment (swese44) [inline src/components/ContainerQuery/ContainerQuery.tsx]: why FocusableComponentProps here?

File: src/components/ContainerQuery/ContainerQuery.test.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
+/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
+import * as React from 'react';
+import ContainerQuery from '.';
+import { create as createRenderer, ReactTestRendererJSON } from 'react-test-renderer';
```
Comment (swese44) [inline src/components/ContainerQuery/ContainerQuery.test.tsx]: Why not use enzyme like the other test files? Just curious

File: src/components/ContainerQuery/ContainerQuery.tsx [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  public componentDidMount() {
+    this.observeElement();
+  }
+
+  public componentWillUnmount() {
+    this.resizeObserver.disconnect();
+  }
+
+  public render() {
+    return <div ref={this.observedElement}>{this.getContent()}</div>;
```
Comment (swese44) [inline src/components/ContainerQuery/ContainerQuery.tsx]: What happens when wide child content is displayed at a wider query and then the viewport is resized smaller than the content can shrink -- if the child content is too wide will we not be able to responsively switch to a smaller query? I wonder if the render method needs to look like this? Might also need some positioning magic so the child wrapper div doesn't push the wrapper too wide?

```
<div>
    <div ref={ref} />
    <div>
        {this.getContent()}
    </div>
</div>
```

Comment (unindented) [reply src/components/ContainerQuery/ContainerQuery.tsx]: Shouldn't we ensure that children are flexible instead?

File: src/components/ContainerQuery/ContainerQuery.tsx (hunk 1 above)
Comment (unindented) [reply src/components/ContainerQuery/ContainerQuery.tsx]: I think it was copy-pasta. I'll fix.

File: src/components/ContainerQuery/ContainerQuery.test.tsx (hunk 2 above)
Comment (unindented) [reply src/components/ContainerQuery/ContainerQuery.test.tsx]: We also seem to be doing this for `Clickable`, not sure why.

File: src/components/ContainerQuery/ContainerQuery.tsx (hunk 1 above)
Comment (jessemartin) [reply src/components/ContainerQuery/ContainerQuery.tsx]: Yes, it was. Let's update.

File: src/components/ContainerQuery/ContainerQuery.test.tsx (hunk 2 above)
Comment (jessemartin) [reply src/components/ContainerQuery/ContainerQuery.test.tsx]: Yeah there's no need to be doing this, we should move to enzyme. (This was copied from the clickable test)

**Comment by jessemartin** [comment]
@unindented `SizeObserver` or `ResizeObserver` make sense to me.

File: src/components/SizeObserver/SizeObserver.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    }
+    if (maxWidth !== undefined && width > maxWidth) {
+      continue;
+    }
+    if (minWidth !== undefined && width < minWidth) {
+      continue;
+    }
+
+    return key;
+  }
```
Comment (chrismohr) [inline src/components/SizeObserver/SizeObserver.tsx]: It took me a little while to grok this with the `continue`s, how would you feel about this refactor?
```
const getMatchingQuerySize = (query: SizeQueriesMap, height: number, width: number) => {
  for (const key of Object.keys(query)) {
    const {
      maxHeight = Number.MAX_SAFE_INTEGER,
      maxWidth = Number.MAX_SAFE_INTEGER,
      minHeight = 0,
      minWidth = 0,
    } = query[key];
    if (height > maxHeight && height < minHeight && width > maxWidth && width < minWidth) {
      return key;
    }
  }
};
```

Comment (jessemartin) [reply src/components/SizeObserver/SizeObserver.tsx]: I agree this is more understandable, thanks! We will update.

File: src/components/SizeObserver/SizeObserver.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
+### Notes for use
+
+SizeObserver allows for conditional rendering based on the component's height or width.
+
+
+### Examples
+
+```js { "props": { "query": { "small": { "maxWidth": 500 }, "large": { "minWidth": 501 } } } }
```
Comment (chrismohr) [inline src/components/SizeObserver/SizeObserver.md]: For what reason are these props are here?

Could it not just be ```js ?

File: src/components/ChoiceGroup/ChoiceGroup.test.tsx [hunk 6]
```diff
@@ -7,7 +7,7 @@ import { IChoiceGroupOptionProps } from 'office-ui-fabric-react/lib/components/C
 
 describe('<ChoiceGroup />', () => {
   let component: ShallowWrapper<ChoiceGroupProps>;
-  let labelWithSublabel: JSX.Element;
```
Comment (chrismohr) [inline src/components/ChoiceGroup/ChoiceGroup.test.tsx]: Mostly for my learning, what caused you to replace `JSX.Element`s?

File: src/components/SizeObserver/SizeObserver.test.tsx [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        case 'large':
+          return <span>Large</span>;
+        case 'small':
+          return <span>Small</span>;
+        default:
+          return <span>Other</span>;
+      }
+    });
+
+    component = mount(<SizeObserver query={query} renderAt={renderAt} />);
```
Comment (chrismohr) [inline src/components/SizeObserver/SizeObserver.test.tsx]: Looks to me like you'd get the same coverage, with clearer snapshots by just looking at the `span`.

```    
component = mount(<SizeObserver query={query} renderAt={renderAt} />).find('span');
```

Comment (chrismohr) [reply src/components/SizeObserver/SizeObserver.test.tsx]: Or maybe `find('div')` to get the wapping div too?

File: src/components/ChoiceGroup/ChoiceGroup.test.tsx (hunk 6 above)
Comment (jessemartin) [reply src/components/ChoiceGroup/ChoiceGroup.test.tsx]: @chrismohr 
JSX.Element is an alias to React.ReactElement.
However, React.ReactElement only describes react classes and stateless functional components. It does not include support for `Fragment` or `Portal`.
React.ReactNode extends React.ReactElement and includes support for Fragments and Portals.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L151

I think @unindented made these changes so our types are explicit and support Fragments.

Comment (chrismohr) [reply src/components/ChoiceGroup/ChoiceGroup.test.tsx]: Cool.  Thanks for the explanation.

File: src/components/SizeObserver/SizeObserver.test.tsx (hunk 7 above)
Comment (jessemartin) [reply src/components/SizeObserver/SizeObserver.test.tsx]: Agreed. I will update.

File: src/components/SizeObserver/SizeObserver.md (hunk 5 above)
Comment (jessemartin) [reply src/components/SizeObserver/SizeObserver.md]: I see we declare the props like this in all of the component readme's I've looked at. I'm not sure why, though. @unindented  @swese44 @richardtagger are these used for anything?

File: src/components/SizeObserver/__snapshots__/SizeObserver.test.tsx.snap [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+exports[`<SizeObserver /> renders at the default size on first render 1`] = `
+<span>
+  Other
+</span>
+`;
+
+exports[`<SizeObserver /> when its size changes renders at the corresponding size 1`] = `
+<span>
+  Other
```
Comment (chrismohr) [inline src/components/SizeObserver/__snapshots__/SizeObserver.test.tsx.snap]: Is this the expected result?  Looking at the test, I would have expected `small`, but maybe I'm misunderstanding.

File: src/components/SizeObserver/__snapshots__/SizeObserver.test.tsx.snap [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+</span>
+`;
+
+exports[`<SizeObserver /> when its size changes twice but both changes match the same query renders at the corresponding size 1`] = `
+<span>
+  Other
```
Comment (chrismohr) [inline src/components/SizeObserver/__snapshots__/SizeObserver.test.tsx.snap]: ... and `Large` here?

File: src/components/SizeObserver/SizeObserver.md (hunk 5 above)
Comment (chrismohr) [reply src/components/SizeObserver/SizeObserver.md]: @jessemartin The other examples use this to add data attributes that are consumed by our visual diff library.  

For example, on Box, 
```
```js { "props": { "data-description": "basic" } }
```

The filename for the visual diff will include "basic" as the name of the example.

**Pull request merged**